rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,592 @@
+    def default_resource_request(cls, config):
+        cf = dict(cls.get_default_config(), **config)
+
+        num_policies = len(cf["max_num_policies_to_train"])
+        if cf["num_gpus"]:
+            num_learner_shards = min(cf["num_gpus"], num_policies)
+            num_gpus_per_shard = cf["num_gpus"] / num_learner_shards
+        else:
+            num_learner_shards = cf.get("num_replay_buffer_shards", 1)
+            num_gpus_per_shard = 0
+
+        eval_config = cf["evaluation_config"]
+
+        # Return PlacementGroupFactory containing all needed resources
+        # (already properly defined as device bundles).
+        return PlacementGroupFactory(
+            bundles=[{
+                # Driver (no GPUs).
+                "CPU": cf["num_cpus_for_driver"],
+            }] + [
+                {
+                    # RolloutWorkers (no GPUs).
+                    "CPU": cf["num_cpus_per_worker"],
+                } for _ in range(cf["num_workers"])
+            ] + [
+                {
+                    # Policy learners (and Replay buffer shards).
+                    "CPU": 1,
+                    "GPU": num_gpus_per_shard,
+                } for _ in range(num_learner_shards)
+            ] + ([
+                {
+                    # Evaluation (remote) workers.
+                    # Note: The local eval worker is located on the driver
+                    # CPU or not even created iff >0 eval workers.
+                    "CPU": eval_config.get("num_cpus_per_worker",
+                                           cf["num_cpus_per_worker"]),
+                    "GPU": eval_config.get("num_gpus_per_worker",
+                                           cf["num_gpus_per_worker"]),
+                } for _ in range(cf["evaluation_num_workers"])
+            ] if cf["evaluation_interval"] else []),
+            strategy=config.get("placement_strategy", "PACK"))
+
+    @classmethod
+    @override(appo.APPOTrainer)
+    def get_default_config(cls) -> TrainerConfigDict:
+        return DEFAULT_CONFIG
+
+    @override(appo.APPOTrainer)
+    def setup(self, config: PartialTrainerConfigDict):
+        # Call super's setup to validate config, create RolloutWorkers
+        # (train and eval), etc..
+        super().setup(config)
+
+        # AlphaStar specific setup:
+        # - Create n policy learner actors (@ray.remote-converted Policies) on
+        #   one or more GPU nodes.
+        # - On each such node, also locate one replay buffer shard.
+        # -
+
+        ma_cfg = self.config["multiagent"]
+        max_num_policies_to_train = self.config["max_num_policies_to_train"]
+
+        # Single CPU replay shard (co-located with GPUs so we can place the
+        # policies on the same machine(s)).
+        ReplayActor = ray.remote(
+            num_cpus=1,
+            num_gpus=0.01
+            if (self.config["num_gpus"] and not self.config["_fake_gpus"]) else
+            0)(MixInMultiAgentReplayBuffer)
```

Comment:
should DistributedLearners handle this? why do we need to pass it in.
I am also not sure why we hardcode num_gpus=0.01 here.
is AlphaStar a GPU only agent, would be nice if we can debug it on our laptop?

Reply:
AlphaStar works with 0 GPUs (see 4x-cartpole learning test added to BUILD) as well as with the _fake_gpu setting (see active compilation test case).

Reply:
The 0.01 is to force the actor to be placed on a GPU node (as it's created before the n policy actors).